File: C++/Project Euler #023.cpp
/* Revisar
 * Solución para el problema #23 de Project Euler
 * "Encuentra la suma de todos los números enteros positivos que no pueden escribirse como suma de dos números abundantes."
 * https://projecteuler.net/problem=23 */

#include <iostream>
#include <math.h>  
using namespace std;

bool isPrime(int n){ //Función para determinar si un numero es primo.
    if (n == 1)return false;
    else if (n < 4)return true;
    else if (n % 2 == 0)return false;
    else if (n < 9)return true;
    else if (n % 3 == 0)return false;
    else{
		double r = sqrt(n); 
		int f = 5;         
        while (f <= r){
            if ((n % f) == 0) return false;
            else if (n % (f + 2) == 0) return false;
            else f += 6;                
        }return true;
    }return true;
}

bool isAbundant(int n){ //Función para determinar si el número es abundante
     if (n < 12) return false; //No hay números abundantes menores de 12
     else if (n % 2 != 0 ) return false; //No hay números abundantes primos ni impares
     else{
          int divisor = 1, total = 0;
          while (divisor <= (n / 2)){              
              if (n%divisor == 0) total += divisor;              
              divisor += 1;
          }if (total > n) return true;
          else if (total <= n) return false;
     }
}

int main (){
    int n, array_size;
    array_size = 28123, n = 0 ;
	int abundant[array_size];
    for (int i = 0; i < array_size; i++){ //Se llena el arreglo de números abundantes
        if (isAbundant(i+1) == true){
           abundant[n] = i+1;
           //cout << abundant[n] << "  "<< n << endl;
           n++;
        }
    }
    array_size = n; //El tamaño del arreglo se reduce a la cantidad de n números abundantes
    int sum, respuesta = 0, menor = 0, mayor = 0;
    for (int number = 1; number <= 28123*2; number++){
        sum = abundant[menor] + abundant[mayor];
        while (number != sum){ //Mientras el número sea diferente a la suma
              if (mayor >= n){
                    mayor = 0;
                    menor ++; //Se vuelven a sumar los números
                    sum = abundant[menor] + abundant[mayor];               
              }if (number != sum){
                 mayor++;
                 sum = abundant[menor] + abundant[mayor];
                 //cout << sum << endl;
                 if (abundant[menor] > abundant[mayor]){
                         respuesta += number, number++;
                         cout << respuesta << endl;
                         mayor = 0, menor = 0;  
                 }         
              }
        }     
    }cout << "La respuesta es " << respuesta << endl;
    return 0;
}

//Suma todos los números impares menores de 28123*2, por propiedad ninguno de estos números es la suma de números abundantes

File: Visual Basic .NET/Project Euler #023.vb
Module Module1
    
    ' Solución para el problema #23 de Project Euler
    ' "¿Cuál es el total de todas las puntuaciones de nombres del archivo?"
    ' https://projecteuler.net/problem=23
    ' Código espagueti de 2016; optimizar

    Function isPrime(ByVal n As Integer) As Boolean 'Función para determinar si un numero es primo.
        If n = 1 Then
            Return False
        ElseIf n < 4 Then
            Return True
        ElseIf n Mod 2 = 0 Then
            Return False
        ElseIf n < 9 Then
            Return True
        ElseIf n Mod 3 = 0 Then
            Return False
        Else
            Dim r As Integer = Math.Floor(Math.Sqrt(n))
            Dim f As Integer = 5
            Do While f <= r
                If n Mod f = 0 Then
                    Return False
                ElseIf n Mod (f + 2) = 0 Then
                    Return False
                Else
                    f += 6
                End If
            Loop
            Return True
        End If
    End Function
    Function isAbundant(ByVal n As Integer) As Boolean 'Función para determinar si el número es abundante
        If n < 12 Then
            Return False
        ElseIf isPrime(n) Then
            Return False
        Else
            Dim divisor As Integer = Nothing
            Dim total As Integer = Nothing
            Do While divisor <= (n / 2)
                divisor += 1
                If n Mod divisor = 0 Then
                    total += divisor
                End If
            Loop
            If total > n Then
                Return True
            ElseIf total <= n Then
                Return False
            End If
            Return True
        End If
    End Function
    Function isDeficient(ByVal n As Integer) As Boolean 'Función para determinar si el número es deficiente
        If n < 12 And n <> 6 Then
            Return True
        ElseIf isPrime(n) Then
            Return True
        Else
            Dim divisor As Integer = Nothing
            Dim total As Integer = Nothing
            Do While divisor <= (n / 2)
                divisor += 1
                If n Mod divisor = 0 Then
                    total += divisor
                End If
            Loop
            If total < n Then
                Return True
            ElseIf total >= n Then
                Return False
            End If
            Return True
        End If
    End Function
    Function isPerfect(ByVal n As Integer) As Boolean 'Función para determinar si el número es perfecto
        If n = 6 Then
            Return True
        End If
        If n < 12 Then
            Return False
        ElseIf isPrime(n) Then
            Return False
        Else
            Dim divisor As Integer = Nothing
            Dim total As Integer = Nothing
            Do While divisor <= (n / 2)
                divisor += 1
                If n Mod divisor = 0 Then
                    total += divisor
                End If
            Loop
            If total = n Then
                Return True
            Else
                Return False
            End If
            Return True
        End If
    End Function

    Sub Main()
        Dim respuesta As Integer = Nothing
        Dim n As Integer = Nothing
        Dim n2 As Integer = Nothing
        Dim n3 As Integer = Nothing
        Do Until n2 > (28123 / 2) + 1
            Do Until isAbundant(n3)
                n3 += 1
            Loop
            If (n2 + n3) > (28123 / 2) + 1 Then
                n2 += 1
                Do Until isAbundant(n2)
                    n2 += 1
                Loop
                n3 = n2
            ElseIf isAbundant(n2) And isAbundant(n3) Then
                respuesta += (n2 + n3)
            End If
            Console.WriteLine(n & "   " & n2 & "   " & n3 & "   " & respuesta)
            'Console.ReadLine()
            n3 += 1
        Loop
        Console.WriteLine("La respuesta es " & respuesta) 'Idk what I'm doing wrong
        Console.ReadLine()
    End Sub
End Module
